File: articles/calculus-for-systems-modeling/articles/line-integrals-and-paths-through-space/fortran/line_integral_audit.f90
program line_integral_audit
  implicit none
  real(8), parameter :: pi = 3.14159265358979323846d0
  print '(A)', 'scenario time_step point_count path_length scalar_line_integral vector_line_integral average_alignment maximum_segment_length path_description warning'
  call audit(1.0d0, 'coarse_path')
  call audit(0.5d0, 'medium_path')
  call audit(0.25d0, 'fine_path')
contains
  subroutine path_point(t, x, y)
    real(8), intent(in) :: t
    real(8), intent(out) :: x, y
    x = t
    y = sin(t)
  end subroutine

  real(8) function scalar_field(x,y)
    real(8), intent(in) :: x,y
    scalar_field = 1.0d0 + y*y
  end function

  subroutine vector_field(x,y,vx,vy)
    real(8), intent(in) :: x,y
    real(8), intent(out) :: vx,vy
    vx = 1.0d0
    vy = x
  end subroutine

  real(8) function dist(x1,y1,x2,y2)
    real(8), intent(in) :: x1,y1,x2,y2
    dist = sqrt((x2-x1)**2 + (y2-y1)**2)
  end function

  subroutine audit(step, scenario)
    real(8), intent(in) :: step
    character(len=*), intent(in) :: scenario
    integer :: i, count
    real(8) :: x1,y1,x2,y2,dx,dy,vx,vy,seg,term,path_len,scalar_total,vector_total,align_sum,max_seg
    character(len=32) :: warning
    count = int((2.0d0*pi)/step) + 1
    path_len = 0.0d0
    scalar_total = 0.0d0
    vector_total = 0.0d0
    align_sum = 0.0d0
    max_seg = 0.0d0
    do i = 0, count-2
      call path_point(i*step, x1, y1)
      call path_point((i+1)*step, x2, y2)
      dx = x2 - x1
      dy = y2 - y1
      seg = dist(x1,y1,x2,y2)
      call vector_field(x1,y1,vx,vy)
      term = vx*dx + vy*dy
      path_len = path_len + seg
      scalar_total = scalar_total + scalar_field(x1,y1) * seg
      vector_total = vector_total + term
      align_sum = align_sum + term / max(seg, 1.0d-12)
      max_seg = max(max_seg, seg)
    end do
    if (step > 0.5d0) then
      warning = 'coarse_path'
    else
      warning = 'synthetic_line_integral'
    end if
    print '(A,1X,F8.3,1X,I8,5F14.6,1X,A,1X,A)', trim(scenario), step, count, path_len, scalar_total, vector_total, align_sum/(count-1), max_seg, 'path', trim(warning)
  end subroutine
end program

File: articles/calculus-for-systems-modeling/articles/line-integrals-and-paths-through-space/cpp/line_integral_audit.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>
#include <utility>

constexpr double PI = 3.14159265358979323846;

std::pair<double,double> path_point(double t){ return {t, std::sin(t)}; }
double scalar_field(double x, double y){ (void)x; return 1.0 + y*y; }
std::pair<double,double> vector_field(double x, double y){ (void)y; return {1.0, x}; }
double distance_between(std::pair<double,double> p, std::pair<double,double> q){ return std::sqrt((q.first-p.first)*(q.first-p.first)+(q.second-p.second)*(q.second-p.second)); }
double dot(std::pair<double,double> a, std::pair<double,double> b){ return a.first*b.first + a.second*b.second; }

void audit(double step, const std::string& scenario){
  int count = static_cast<int>((2.0*PI)/step) + 1;
  double path_len = 0.0, scalar_total = 0.0, vector_total = 0.0, align_sum = 0.0, max_seg = 0.0;
  for(int i=0; i<count-1; i++){
    double t = i * step;
    auto p = path_point(t);
    auto q = path_point(t+step);
    auto disp = std::make_pair(q.first-p.first, q.second-p.second);
    double seg = distance_between(p,q);
    auto field = vector_field(p.first,p.second);
    double term = dot(field, disp);
    path_len += seg;
    scalar_total += scalar_field(p.first,p.second) * seg;
    vector_total += term;
    align_sum += term / std::max(seg, 1e-12);
    max_seg = std::max(max_seg, seg);
  }
  std::string warning = step > 0.5 ? "Time step is coarse; path turns and field variation may be undersampled." : "Synthetic line-integral audit; document path field units and interpolation.";
  std::cout << scenario << "," << step << "," << count << "," << path_len << "," << scalar_total << "," << vector_total << "," << align_sum/(count-1) << "," << max_seg << ",path r(t)=<t,sin(t)>," << warning << "\n";
}

int main(){
  std::cout << std::fixed << std::setprecision(12);
  std::cout << "scenario,time_step,point_count,path_length,scalar_line_integral,vector_line_integral,average_alignment,maximum_segment_length,path_description,warning\n";
  audit(1.0, "coarse_path");
  audit(0.5, "medium_path");
  audit(0.25, "fine_path");
}
